src/lean_dojo/data_extraction/traced_data.py

```diff
@@ -1229,19 +1230,32 @@ def _save_xml_to_disk(tf: TracedFile) -> None:
         oup.write(tf.to_xml())
 
 
-def _build_dependency_graph(traced_files: List[TracedFile]) -> nx.DiGraph:
+def _build_dependency_graph(
+    seed_files: List[TracedFile], root_dir: Path, repo: LeanGitRepo
+) -> nx.DiGraph:
     G = nx.DiGraph()
 
-    for tf in traced_files:
+    for tf in seed_files:
         tf_path_str = str(tf.path)
         assert not G.has_node(tf_path_str)
         G.add_node(tf_path_str, traced_file=tf)
 
-    for tf in traced_files:
+    traced_files = seed_files.copy()
+    i = 0
+
+    while i < len(traced_files):
+        tf = traced_files[i]
         tf_path_str = str(tf.path)
+
         for dep_path in tf.get_direct_dependencies():
             dep_path_str = str(dep_path)
-            assert G.has_node(dep_path_str)
+            if not G.has_node(dep_path_str):
+                xml_path = to_xml_path(root_dir, dep_path, repo.uses_lean4)
+                tf_dep = TracedFile.from_xml(root_dir, xml_path, repo)
+                assert tf_dep.path == tf.path
+                G.add_node(dep_path_str, traced_file=tf_dep)
+                traced_files.append(tf_dep)
```

Comment:
Maybe you wanna drop a comment here on why a second-order dependency is needed. :)

Reply:
If you have `import A` in the current file, and the file A has `import B`, then you should be able to use everything defined in B in the current file?

Reply:
Are you stating it as a question since it is obvious, or because you are unsure?
Why do we need to use everything that B imports? Are the objects in B used later on to construct training data for the original file, or only objects from A. 
For the usecase that I am imaging, my training data would only reference to objects in A. (but I am still prototyping).

Reply:
You are right that an Import of A, where A imports B as well, makes the definitions of A and B available in the original file. So your construction is technically correct and probably we should merge it.

But this implies that also all the std, lean4, docs, and other library dependencies are read and imported, which does not provide any value for us but costs a lot of time. 
Not sure what to do about it. For my local developments, I think I will these imports off.

Reply:
This will not necessarily import all files. It includes files that are imported (directly or indirectly) by the target repo. But many other files are not imported at all.

Reply:
Yeah, you are right. But for example, the mathlib4/Mathlib/Algebra/Field/Defs.lean also imports definitions from the std package: import Std.Data.Rat
And I think we don't need these to parse other libraries (like std) as well.

Reply:
We would need them to figure out the set of all accessible premises for the premise retrieval. But feel free to disable them locally.

Reply:
Ah, its the special case of premise retrieval. I see. Thanks.